src/get-workspaces-paths.js

```diff
@@ -31,51 +35,148 @@ function processYarn({ stdout }) {
   return workspaces;
 }
 
+function processGlobs(cwd, _2dFilesArray){
+  let _1dFilesArray = Array.prototype.concat.apply([], _2dFilesArray);
+
+  let uniqueFiles = [...new Set(_1dFilesArray)];
+
+  let packagePaths = uniqueFiles.map(file => path.join(cwd, file, 'package.json'));
+
+  function prunePackagePaths(packagePath) {
+    let index = packagePaths.indexOf(packagePath);
+
+    if (index > -1) {
+      packagePaths.splice(index, 1);
+    }
+  }
+
+  for (let packagePath of packagePaths) {
+    if (fs.existsSync(packagePath)) {
+      let packageJson = readJsonSync(packagePath);
+
+      let neededKeys = ['name', 'version'];
+
+      // not valid package.json if name and version are missing
+      if (!neededKeys.every(key => Object.keys(packageJson).includes(key))){
+        prunePackagePaths(packagePath);
+      }
+    } else {
+      prunePackagePaths(packagePath);
+    }
+  }
+
+  let workspaces = packagePaths.map(workspace => path.relative(cwd, workspace).replace('/package.json', ''));
+
+  return workspaces;
+}
+
 async function getWorkspacesPaths({
   cwd,
+  shouldSpawn,
 }) {
   let workspacePackageJson = await readJson(path.join(cwd, 'package.json'));
 
   let { workspaces } = workspacePackageJson;
 
+  let packagesGlobs;
+
+  let spawn = shouldSpawn || false;
+
   if (!workspaces) {
-    workspaces = processPnpm(
-      await execa('pnpm', ['recursive', 'exec', '--', 'node', '-e', 'console.log(process.cwd())'], {
+    if (spawn){
+      workspaces = processPnpm(
+        await execa('pnpm', ['recursive', 'exec', '--', 'node', '-e', 'console.log(process.cwd())'], {
+          cwd,
+        }),
         cwd,
-      }),
-      cwd,
-    );
+      );
+    } else {
+      // read packagesGlobs excluding packagesGLobs starting with !
+      // https://pnpm.io/pnpm-workspace_yaml
+      packagesGlobs = jsYaml.load(
```

Comment:
This could be `packagesGlobs = processPnpmYaml(await fs.readFile(path.join(cwd, 'pnpm-workspace.yaml')));` so it can be shared.